Sources/Implementation/DefaultDatafileHandler.swift

```diff
@@ -255,56 +257,57 @@ class DefaultDatafileHandler: OPTDatafileHandler {
-            
-            let fileURL = dir.appendingPathComponent(sdkKey, isDirectory: false)
-            
-            //writing
-            do {
-                try dataFile.write(to: fileURL, options: .atomic)
-            } catch {/* error handling here */
-                logger.e("Problem saving datafile for key " + sdkKey)
-            }
+    func getDatafileCache(sdkKey:String) -> OPTDataStore {
+        if let cache = datafileCache[sdkKey] {
+            return cache
+        }
+        else {
+            #if os(tvOS)
+            let store = DataStoreUserDefaults()
+            datafileCache[sdkKey] = store
+            return store
+            #else
+            let store = DataStoreFile<Data>(storeName: sdkKey)
+            datafileCache[sdkKey] = store
+            return store
+            #endif
         }
     }
     
+    func saveDatafile(sdkKey: String, dataFile: Data) {
+        getDatafileCache(sdkKey: sdkKey).saveItem(forKey: sdkKey, value: dataFile)
+    }
+    
     func loadSavedDatafile(sdkKey: String) -> Data? {
-        if let dir = FileManager.default.urls(for: .documentDirectory, in: .userDomainMask).first {
-            
-            let fileURL = dir.appendingPathComponent(sdkKey)
-            
-            //reading
-            do {
-                let data = try Data(contentsOf: fileURL)
-                return data
-            } catch {/* error handling here */
-                logger.e("Problem loading datafile for key " + sdkKey)
-            }
-        }
-        
-        return nil
+        return getDatafileCache(sdkKey: sdkKey).getItem(forKey: sdkKey) as? Data
     }
     
     func isDatafileSaved(sdkKey: String) -> Bool {
-        if let dir = FileManager.default.urls(for: .documentDirectory, in: .userDomainMask).first {
-            let fileURL = dir.appendingPathComponent(sdkKey)
-            return FileManager.default.fileExists(atPath: fileURL.path)
-        }
-        
-        return false
+        let item = getDatafileCache(sdkKey: sdkKey).getItem(forKey: sdkKey)
+        return item != nil && (item as? Data) != nil
     }
     
     func removeSavedDatafile(sdkKey: String) {
-        if let dir = FileManager.default.urls(for: .documentDirectory, in: .userDomainMask).first {
-            let fileURL = dir.appendingPathComponent(sdkKey)
-            if FileManager.default.fileExists(atPath: fileURL.path) {
-                try? FileManager.default.removeItem(at: fileURL)
-            }
+        getDatafileCache(sdkKey: sdkKey).removeItem(sdkKey: sdkKey)
+    }
+}
+
+extension OPTDataStore {
+    public func removeItem(sdkKey:String) {
+        if let store = self as? DataStoreUserDefaults {
+            store.removeItem(sdkKey: sdkKey)
+        }
+        else if let store = self as? DataStoreFile<Data> {
+            store.removeItem(sdkKey: sdkKey)
+        }
+        else if let store = self as? DataStoreMemory<Data> {
+            store.removeItem(sdkKey: sdkKey)
```

Comment:
It must be for backward compatibility. Can't we leave it empty? All our datastores override it. Also we can move it to the end of OPTDataStore for readability.

public func removeItem(sdkKey: String) { }

Reply:
no because that is the method called on the extension (i.e. the other implementation is not an override since it is an extension).  the only alternative is to add it to the data store protocol directly.